### DISCUSSION TITLE
Issues mapping some some collection properties and dates with IQueryable

### DISCUSSION THREAD

**Discussion by Atulin** [riok/mapperly#1380] [Q&A] [answered]
Source:
```cs
public class Story
{
	// ...
	public DateTime CreationDate { get; set; }
	public DateTime? PublicationDate { get; set; }
	public IList<Chapter> Chapters { get; set; }
}

public class Chapter
{
	// ...
	public CommentsThread CommentsThread { get; set; }
}

public class CommentsThread 
{
	// ...
	public int CommentsCount { get; set; }
}
```

Target:
```cs
public class StoryDetails
{
	public DateTime ReleaseDate { get; set; }
	public bool IsPublished { get; set; }
	public bool CommentsCount { get; set; }
}
```

Mapper:
```cs
[Mapper]
public static partial class StoryMapper
{
	public static partial IQueryable<StoryDetails> ProjectToDetails(this IQueryable<Story> q);

	// Does not work
	private static int MapCommentsCount(IList<Chapter> chapters)
		=> chapters.Sum(c => c.CommentsThread.CommentsCount);

	// Neither do those two
	private static DateTime MapReleaseDate(Story story) => story.PublicationDate ?? story.CreationDate;
	private static bool MapIsPublished(DateTime? publicationDate) => publicationDate != null;
}
```

![image](https://github.com/riok/mapperly/assets/11233299/fd847b69-3e11-4752-8c58-b6e4d86b0b6b)

**Comment by latonz** [comment]
Is the posted target class exactly as it is in your code? The error message indicates that the members are required, which is not the case in the posted code.
Anyway, Mapperly does not know it should map `Story.Chapters` to `StoryDetails.CommentsCount`. The same for the other properties. You can use [`MapProperty`](https://mapperly.riok.app/docs/configuration/mapper/#properties--fields) in combination with `MapProperty.Use`, see [docs](https://mapperly.riok.app/docs/configuration/mapper/#user-implemented-property-mappings) or before and after map ([docs](https://mapperly.riok.app/docs/configuration/before-after-map/)).

**Reply by Atulin** [reply]
The classes do have [more properties](https://github.com/Genfic/Ogma/tree/master/Ogma3%2FData%2FStories) in them, yes.

I tried using the `Use` but to no effect, it generated no new code. I'll give it another shot and report back.

**Reply by latonz** [reply, marked as answer]
Note: you need to apply the `MapProperty` to a mapping method with the classes as parameter / return types, not wrapped in `IQueryable`. Otherwise Mapperly tries to apply the `MapProperty` attribute to the `IQueryable` which doesn't have these properties.

```csharp
public static partial IQueryable<StoryDetails> ProjectToDetails(this IQueryable<Story> q);

[MapProperty(...)]
private static partial StoryDetails Map(Story source);
```

Mapperly will pick up the configuration from the `Map` method whenever mapping from `Story` to `StoryDetails`.
See also the [documentation](https://mapperly.riok.app/docs/configuration/queryable-projections/#property-configurations).

**Reply by Atulin** [reply]
Ah, that was probably my mistake, then! I was adding `[MapProperty]` to the `IQueryable` mapping
